### PR TITLE
Feature/advertisement

### PR DESCRIPTION
**Pull request by rumbarum**
Create 1 view class, Modify variable name …
1. AdvertisementCreateGetListView
    method: POST
        creating advertisement by authroized advertiser with interests(fixed) and
        tags(which is not fixed)

    method: GET
        delivering advertisement list with id, title, thumbnail,
        price_per_view and created_at by clients request querystring condition
    -condition:
        offset         = [integer](default=0),
        limit          = [integer](default=3),
        category_id    = [integer],
        price_per_view = [string],
        like           = [string],
        interests      = [string],
        advertiser     = [string],
        tag            = [string]

        Price_per_view, like's querystring should be 'set' or will return Error.
        Interests, advertiser work with 'set' requiring user login, advertiser login.
        Tag can get any string.

2. variable reqst changed to request in UserSignupView

3. Conditional Filter added
AdvertisementCreateGetListView deliver only
Advertisemnet.objects.filter(deleted=False) and deliver Advertisemnet.objects on condition of switch.

### CODE REVIEW

File: advertisement/views.py [hunk 1]
```diff
@@ -1,3 +1,152 @@
...
-# Create your views here.
+from django.http            import HttpResponse, JsonResponse
+from django.views           import View
+from django.core.exceptions import ValidationError
+
+from user.utils  import advertiser_login_required, user_login_required, advertiser_login_mark, user_login_mark
+from user.models import User, Advertiser, InterestsType, State, Gender, UsersInterests
+from .models     import (AdvertisementCategory,Tag,  AdvertisementTag, Advertisement,
+                         AdvertisementLike, QuizHistory, AdvertisementsInterests 
+                        )
```
Comment (rampart81) [inline advertisement/views.py]: ===>

```
from .models  import (
    AdvertisementCategory,Tag,  
    AdvertisementTag, 
    Advertisement,
    AdvertisementLike, 
    QuizHistory, 
    AdvertisementsInterests 
)
```

File: advertisement/views.py [hunk 2]
```diff
@@ -1,3 +1,152 @@
...
+
+    @user_login_mark
+    @advertiser_login_mark
+    def get(self, request):        
+        advertisements = Advertisement.objects.filter(deleted=False) 
+    
+        try:
+            offset         = int(request.GET.get('offset', '0'))
+            limit          = int(request.GET.get('limit', '3'))
+            category_id    = int(request.GET.get('category_id','0'))
```
Comment (rampart81) [inline advertisement/views.py]: `category_id = int(request.GET['category_id']) if 'category_id' in request.GET else None`

File: advertisement/views.py [hunk 3]
```diff
@@ -1,3 +1,152 @@
...
+            price_per_view = request.GET.get('price_per_view', None)
+            view           = request.GET.get('view', None)
+            like           = request.GET.get('like', None)
+            interests      = request.GET.get('interests', None)
+            tag            = request.GET.get('tag', None)
+            advertiser     = request.GET.get('advertiser', None)
+            
+            if advertiser == 'set':
```
Comment (rampart81) [inline advertisement/views.py]: `if hasattr(request, 'advertiser'):`

File: advertisement/views.py [hunk 4]
```diff
@@ -1,3 +1,152 @@
...
+                        "off_advertisement":querytolist(off_advertisements)}, 
+                        status=200
+                        )
+
+            advertisements = advertisements.filter(switch=True)
+
+            if category_id != 0:
+                advertisements = advertisements.filter(ad_category_id=category_id)       
+            
+            if price_per_view == 'set':
```
Comment (rampart81) [inline advertisement/views.py]: `if price_per_view:`

Comment (rampart81) [reply advertisement/views.py]:
```
Reuqest Body
{
    "order_by" : [
        {"field" : "price_per_view", "asc" :  False},
        { "field" : "like_count, "asc": False}
    ]
}

order_by_fields = data["order_by"]
asc = data["asc"]

for order_by in order_by_fields:
    field = order_by["field"]
    asc  = "order_by["asc"]
    order_direction = "-" if not asc else ""
    advertiserments.order_by(f"{order_direction}{field}")
```

File: advertisement/views.py [hunk 5]
```diff
@@ -1,3 +1,152 @@
...
+                advertisements = advertisements.order_by('-view_count')
+            
+            if tag: 
+                tag = Tag.objects.filter(name=tag)[0]
+                advertisements = advertisements.filter(tag = tag)
+            
+            if interests == 'set': 
+                interests_list = request.user.interests.all()
+                advertisements = (
+                    list(set(advertisements.filter(interests__in=interests_list)))
```
Comment (rampart81) [inline advertisement/views.py]: `distinct`

Comment (rampart81) [reply advertisement/views.py]: `count`

File: advertisement/views.py [hunk 6]
```diff
@@ -1,3 +1,152 @@
...
+                ) 
+
+            if  not(advertiser or price_per_view or like or tag or view or interests): 
+                advertisements = advertisements.order_by('-created_at')
+
+            if type(advertisements) == list:
```
Comment (rampart81) [inline advertisement/views.py]: 이런 식의 타입으로 체크는 지양 합니다

File: user/utils.py [hunk 7]
```diff
@@ -40,14 +40,60 @@ def decorated_function(self, request, *args, **kwargs):
...
             except jwt.DecodeError:
                 return JsonResponse({"ERROR":"INVALID_TOKEN"}, status=401)
             except Advertiser.DoesNotExist:
+
                 return JsonResponse({"ERROR":"ID_NOT_EXIST"}, status=401)
         else:
             return JsonResponse({"ERROR":"LOGIN_REQUIERD"}, status=401)
     return decorated_function
 
+def user_login_mark(func):
```
Comment (rampart81) [inline user/utils.py]: `user_login_mark` ==> `optional_user_login`

File: advertisement/views.py (hunk 4 above)
Comment (rumbarum) [reply advertisement/views.py]:
```
```

**Pull request closed without merging**